publish-mqtt/src/main.rs

```diff
@@ -118,99 +106,40 @@ async fn main() -> Result<(), Box<dyn Error + Send + Sync>> {
     print_sensors(&sensors, &sensor_names);
-    let connected_sensors = connect_sensors(&sensors);
-
-    let mut nodes = vec![];
+    let (named_sensors, unnamed_sensors): (Vec<_>, Vec<_>) = sensors
+        .into_iter()
+        .partition(|sensor| sensor_names.contains_key(&sensor.get_address().unwrap()));
+    println!("Connecting to named sensors first");
+    let mut connected_sensors = connect_sensors(&named_sensors);
+    println!("Connecting to unnamed sensors");
+    connected_sensors.extend(connect_sensors(&unnamed_sensors));
+
+    let properties = vec![
+        Property::new("temperature", "Temperature", "float", "ºC"),
+        Property::new("humidity", "Humidity", "integer", "%"),
+        Property::new("battery", "Battery level", "integer", "%"),
+    ];
     for sensor in &connected_sensors {
         let mac_address = sensor.get_address()?;
         let node_id = mac_address.replace(":", "");
-        let node_base = format!("{}/{}", device_base, node_id);
         let node_name = sensor_names.get(&mac_address).unwrap_or(&mac_address);
-        nodes.push(node_id);
-        publish_retained(&requests_tx, format!("{}/$name", node_base), node_name).await?;
-        publish_retained(&requests_tx, format!("{}/$type", node_base), "Mijia sensor").await?;
-        publish_retained(
-            &requests_tx,
-            format!("{}/$properties", node_base),
-            "temperature,humidity,battery",
-        )
-        .await?;
-        publish_retained(
-            &requests_tx,
-            format!("{}/temperature/$name", node_base),
-            "Temperature",
-        )
-        .await?;
-        publish_retained(
-            &requests_tx,
-            format!("{}/temperature/$datatype", node_base),
-            "float",
-        )
-        .await?;
-        publish_retained(
-            &requests_tx,
-            format!("{}/temperature/$unit", node_base),
-            "ºC",
-        )
-        .await?;
-        publish_retained(
-            &requests_tx,
-            format!("{}/humidity/$name", node_base),
-            "Humidity",
-        )
-        .await?;
-        publish_retained(
-            &requests_tx,
-            format!("{}/humidity/$datatype", node_base),
-            "integer",
-        )
-        .await?;
-        publish_retained(&requests_tx, format!("{}/humidity/$unit", node_base), "%").await?;
-        publish_retained(
-            &requests_tx,
-            format!("{}/battery/$name", node_base),
-            "Battery level",
-        )
-        .await?;
-        publish_retained(
-            &requests_tx,
-            format!("{}/battery/$datatype", node_base),
-            "integer",
-        )
-        .await?;
-        publish_retained(&requests_tx, format!("{}/battery/$unit", node_base), "%").await?;
+        homie.add_node(Node::new(
+            node_id,
+            node_name.clone(),
+            "Mijia sensor".to_string(),
+            properties.clone(),
```

Comment:
is this clone necessary? Are we using properties anywhere else?

Reply:
Yes, because we are using the same `properties` for all the nodes. (Note that `properties` is defined outside this loop.)